File: go/attic/repl.go
// =============================================================================
// repl.go - REPL Loop with Line Editor, Command Translation, and Assembly Mode
// =============================================================================
//
// This file implements the REPL (Read-Eval-Print Loop) for the Attic CLI.
// The REPL reads user input via the LineEditor (see lineeditor.go), processes
// local dot-commands (mode switching, help, quit), translates mode-specific
// commands into protocol format (see translate.go), and sends them to
// AtticServer via a Unix socket client.
//
// The REPL operates in three modes:
//   - Monitor: 6502 debugging (disassembly, breakpoints, memory inspection)
//   - BASIC:   BASIC program entry and execution
//   - DOS:     Disk image management
//
// Additionally, the REPL supports an interactive assembly sub-mode that is
// entered when the server responds to an "assemble $XXXX" command with
// "ASM $XXXX". In this mode, the prompt changes to "$XXXX: " and each
// line is sent as "asm input <instruction>" until the user enters a blank
// line or "." to exit.
//
// Mode switching is handled locally (no server round-trip). Protocol commands
// are sent as raw text strings via the CLI text protocol.
//
// =============================================================================

package main

// GO CONCEPT: Buffered I/O (bufio package)
// -----------------------------------------
// Go's "bufio" package provides buffered readers and scanners for efficient
// I/O. Raw os.Stdin reads one byte at a time (expensive system calls);
// bufio wraps it with an internal buffer for better performance.
//
// Two common approaches:
//   bufio.NewReader(os.Stdin) — low-level, gives you ReadString('\n'), etc.
//   bufio.NewScanner(os.Stdin) — higher-level, iterates lines with Scan()
//
// Scanner is simpler for line-by-line reading (which is what a REPL needs),
// while Reader is better when you need more control.
//
// In this file, we no longer use bufio directly for REPL input — that's
// been moved to the LineEditor (lineeditor.go). But the concept is still
// relevant for understanding the non-interactive fallback path.
//
// Compare with Python: Python file objects are buffered by default.
// `sys.stdin` can be iterated line by line: `for line in sys.stdin:`.
// For explicit buffering, use `io.BufferedReader`. The `readline` module
// adds line editing (history, tab completion) on top of stdin.
//
// The "strings" package provides string manipulation functions. Go strings
// are immutable (like Swift), so operations return new strings rather than
// modifying in place.
import (
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/attic/atticprotocol"
)

// GO CONCEPT: Custom Types (Type Definitions)
// --------------------------------------------
// "type REPLMode int" creates a NEW named type based on int. This is NOT
// just a type alias — REPLMode and int are different types and can't be
// mixed without explicit conversion.
//
// This is Go's way of creating enums: define a named type, then use
// iota-based constants for the values. It's more manual than Swift enums
// (no associated values, no pattern matching) but gets the job done.
//
// Compare to Swift:
//   enum REPLMode: String { case monitor, basic, dos }
//
// Go's approach:
//   type REPLMode int
//   const ( ModeMonitor REPLMode = iota; ModeBasic; ModeDOS )
//
// Compare with Python: Python uses `enum.IntEnum` for integer enums:
//   `class REPLMode(IntEnum): MONITOR = 0; BASIC = 1; DOS = 2`
// Python enums are full classes with rich features: iteration, names,
// and pattern matching support.

// REPLMode represents the current operating mode of the REPL.
type REPLMode int

// GO CONCEPT: iota — Auto-Incrementing Constants
// ------------------------------------------------
// "iota" is Go's constant generator. Inside a const block, iota starts
// at 0 and increments by 1 for each constant. It resets to 0 in each
// new const block.
//
//   const (
//       ModeMonitor REPLMode = iota  // 0
//       ModeBasic                     // 1 (type and iota+1 carried forward)
//       ModeDOS                       // 2
//   )
//
// The type "REPLMode" and the expression "iota" carry forward to
// subsequent lines that omit them. This is a special Go shorthand for
// const blocks.
//
// You can also do arithmetic with iota:
//   const ( KB = 1 << (10 * (iota + 1)); MB; GB; TB )
//   // KB=1024, MB=1048576, GB=1073741824, TB=1099511627776
//
// Compare to Swift: Swift enums auto-assign raw values similarly:
//   enum Mode: Int { case monitor = 0, basic, dos }
//
// Compare with Python: Python's `enum.auto()` serves the same purpose:
//   `class REPLMode(IntEnum): MONITOR = auto(); BASIC = auto()`
// Values start at 1 by default (not 0), but you can override with a
// custom `_generate_next_value_` method.
const (
	// ModeMonitor is the 6502 debugging mode.
	ModeMonitor REPLMode = iota
	// ModeBasic is the BASIC programming mode.
	ModeBasic
	// ModeDOS is the disk management mode.
	ModeDOS
)

// GO CONCEPT: Methods on Custom Types
// -------------------------------------
// Go doesn't have classes, but you can define methods on ANY named type
// (not just structs). The syntax uses a "receiver" before the function name:
//
//   func (m REPLMode) prompt() string { ... }
//         ^^^^^^^^^^^
//         This is the "receiver" — it's like "self" in Swift.
//
// The receiver binds the method to the type. You call it as:
//   mode := ModeBasic
//   mode.prompt()  // returns "[basic] > "
//
// Receivers can be value receivers (m REPLMode) or pointer receivers
// (m *REPLMode):
//   - Value receiver: gets a COPY of the value (can't modify the original)
//   - Pointer receiver: gets a reference (can modify the original)
//
// Use a value receiver when the method only reads the value (like here).
// Use a pointer receiver when the method needs to modify the value.
//
// Compare to Swift:
//   Swift methods are always implicitly called on "self":
//     func prompt() -> String { switch self { ... } }
//   Go requires you to name the receiver explicitly:
//     func (m REPLMode) prompt() string { switch m { ... } }
//
// Compare with Python: Python methods always take `self` as the first
// parameter (like Go's receiver, but always named `self`):
//   `def prompt(self) -> str: ...`
// Python allows adding methods to any class, including subclasses of
// built-in types.

// prompt returns the display prompt for the current REPL mode.
func (m REPLMode) prompt() string {
	switch m {
	case ModeMonitor:
		return "[monitor] > "
	case ModeBasic:
		return "[basic] > "
	case ModeDOS:
		return "[dos] D1:> "
	default:
		return "> "
	}
}

// GO CONCEPT: Dependency Injection
// ---------------------------------
// runREPL accepts a *LineEditor as a parameter rather than creating one
// internally. This is "dependency injection" — the caller (main.go) creates
// the LineEditor and passes it in. Benefits:
//
//   1. Testability: Tests can pass a mock or test-configured LineEditor.
//   2. Lifetime control: The caller manages creation and cleanup (Close).
//   3. Flexibility: The same REPL code works with different input sources.
//
// This is the same pattern used in the Swift CLI, where the LineEditor
// is created in the main entry point and passed to runSocketREPL().
//
// Compare with Swift: Swift uses initializer injection or property injection:
//   func runSocketREPL(lineEditor: LineEditor, client: CLISocketClient)
//
// Compare with Python: Python uses constructor injection:
//   def run_repl(editor: LineEditor, client: Client) -> None: ...
// Python also supports dependency injection frameworks like `inject` or
// `dependency-injector`, but simple parameter passing is most common.

// runREPL runs the main REPL loop.
//
// It reads user input via the LineEditor, processes local dot-commands
// (mode switching, help, quit), translates mode-specific commands via
// translateToProtocol (see translate.go), and sends them to the server.
// Responses are displayed with multi-line expansion (replacing Record
// Separator characters with newlines).
//
// The REPL supports an interactive assembly sub-mode: when the server
// responds with "ASM $XXXX", the prompt changes to "$XXXX: " and each
// line is sent as "asm input <instruction>" until a blank line or ".".
//
// The REPL exits on:
//   - .quit command
//   - .shutdown command (also stops the server)
//   - EOF (Ctrl-D in interactive mode, end of piped input)
//   - LineEditor read error
func runREPL(client *atticprotocol.Client, editor *LineEditor, atasciiMode bool) {
	mode := ModeBasic

	// GO CONCEPT: State Machine for Assembly Sub-Mode
	// ------------------------------------------------
	// The assembly sub-mode is a state machine embedded in the REPL loop.
	// Two boolean/integer variables track the state:
	//   - inAssemblyMode: whether we're currently in assembly mode
	//   - assemblyAddress: the next address for assembly prompt display
	//
	// When inAssemblyMode is true, the REPL:
	//   - Shows "$XXXX: " prompts instead of mode prompts
	//   - Sends "asm input <instruction>" for each line
	//   - Sends "asm end" on blank line or "." to exit
	//   - Extracts the next address from the server response
	//
	// Compare with Swift: The Swift CLI uses the same approach with
	// `var inAssemblyMode = false` and `var assemblyAddress: UInt16 = 0`.
	//
	// Compare with Python: Python would use instance variables on a
	// REPL class: `self.in_assembly_mode = False`.
	inAssemblyMode := false
	var assemblyAddress uint16

	// GO CONCEPT: Infinite Loops
	// ---------------------------
	// "for { ... }" is Go's infinite loop (equivalent to "while true").
	// We use break/return to exit the loop. Most REPLs use this pattern:
	// loop forever, reading input and processing it, until the user quits.
	//
	// Compare with Python: `while True:` is Python's infinite loop. `break`
	// and `return` exit it, just like in Go.
	for {
		// Determine the prompt based on current state.
		// In assembly mode, show the next address; otherwise show the mode.
		var prompt string
		if inAssemblyMode {
			prompt = fmt.Sprintf("$%04X: ", assemblyAddress)
		} else {
			prompt = mode.prompt()
		}

		// Read a line of input using the LineEditor.
		// In interactive mode, this provides Emacs keybindings, history
		// navigation (up/down arrows, Ctrl-R), and persistent history.
		// In non-interactive mode, it prints the prompt and reads from stdin.
		line, err := editor.GetLine(prompt)
		if err != nil {
			// GO CONCEPT: Comparing Errors with ==
			// --------------------------------------
			// io.EOF is a sentinel error value. We compare with == because
			// it's a unique package-level variable, not a wrapped error.
			// For wrapped errors (created with fmt.Errorf %w), use
			// errors.Is(err, target) instead.
			//
			// Compare with Swift: Swift checks for EOF via nil return:
			//   guard let line = readLine() else { break }
			//
			// Compare with Python: Python catches EOFError:
			//   try: line = input(prompt)
			//   except EOFError: break
			if err == io.EOF {
				// Clean EOF — user pressed Ctrl-D or piped input ended.
				// If in assembly mode, end the session cleanly.
				if inAssemblyMode {
					_, _ = client.SendRaw("asm end")
					inAssemblyMode = false
				}
				fmt.Println()
				return
			}
			// Unexpected error — log and exit.
			fmt.Fprintf(os.Stderr, "Input error: %v\n", err)
			return
		}

		// --- Interactive Assembly Sub-Mode ---
		// When active, user input is routed to "asm input" / "asm end"
		// instead of being interpreted as normal REPL commands.
		if inAssemblyMode {
			trimmed := strings.TrimSpace(line)

			// Empty line or "." exits assembly mode.
			if trimmed == "" || trimmed == "." {
				resp, err := client.SendRaw("asm end")
				if err != nil {
					fmt.Fprintf(os.Stderr, "Error: %v\n", err)
				} else if resp.IsOK() {
					if resp.Data != "" {
						fmt.Println(resp.Data)
					}
				} else {
					fmt.Fprintf(os.Stderr, "Error: %s\n", resp.Data)
				}
				inAssemblyMode = false
				continue
			}

			// Feed instruction to the active assembly session.
			// Response format: "formatted line\x1E$XXXX"
			// The first part is the assembled output to display.
			// The second part (after separator) is the next address.
			resp, err := client.SendRaw("asm input " + trimmed)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error: %v\n", err)
				continue
			}

			if resp.IsOK() {
				// Split on Record Separator to get assembled line + next address.
				parts := strings.SplitN(resp.Data, atticprotocol.MultiLineSeparator, 2)
				// Print the assembled line.
				fmt.Println(parts[0])
				// Extract next address for the prompt.
				if len(parts) > 1 {
					if addr, ok := parseHexAddress(parts[1]); ok {
						assemblyAddress = addr
					}
				}
			} else {
				// Assembly error — session stays alive, user can retry.
				fmt.Fprintf(os.Stderr, "Error: %s\n", resp.Data)
			}
			continue
		}

		// --- Normal Mode ---
		// GO CONCEPT: String Functions
		// ----------------------------
		// Go's strings package provides functions (not methods) for string
		// manipulation. Unlike Swift's "string.trimmingCharacters(in:)",
		// Go uses "strings.TrimSpace(string)".
		//
		// These are standalone functions, not methods, because Go's string
		// type is a built-in primitive. You can't add methods to built-in types.
		//
		// Compare with Python: Python strings have methods directly (not
		// standalone functions): `line.strip()`, `line.startswith(".")`,
		// `line.split()`, `line.upper()`.
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// GO CONCEPT: String Case Conversion for Commands
		// ------------------------------------------------
		// strings.ToLower() converts a string to lowercase. We use this
		// to handle dot-commands case-insensitively — ".Quit", ".QUIT",
		// and ".quit" all work the same way. This matches the user-friendly
		// behavior of the Swift CLI.
		//
		// Compare with Swift: Swift's lowercased() method:
		//   switch line.lowercased() { case ".quit": ... }
		//
		// Compare with Python: Python's lower() method:
		//   match line.lower(): case ".quit": ...
		lowerLine := strings.ToLower(line)

		// Handle local dot-commands (processed by the CLI, not sent to server).
		//
		// GO CONCEPT: Switch on Computed Values
		// --------------------------------------
		// Go's switch can match on any comparable type — here we switch on
		// a lowercase string. Unlike C, Go switch cases don't fall through
		// by default (no break needed). The "continue" at the end of each
		// case skips the server-send logic below and returns to the top of
		// the REPL loop.
		//
		// Compare with Swift: Swift's switch with string patterns.
		// Compare with Python: Python 3.10+ match statement.
		handled := true
		switch lowerLine {
		case ".quit":
			return
		case ".shutdown":
			// .shutdown tells the server to stop, then exits the CLI.
			_, _ = client.SendRaw("shutdown")
			return
		case ".monitor":
			mode = ModeMonitor
			fmt.Println("Switched to Monitor mode")
		case ".basic":
			mode = ModeBasic
			fmt.Println("Switched to BASIC mode")
		case ".dos":
			mode = ModeDOS
			fmt.Println("Switched to DOS mode")
		case ".help":
			printHelp(mode, "")
		default:
			handled = false
		}

		if handled {
			continue
		}

		// Handle .help with a topic argument.
		if strings.HasPrefix(lowerLine, ".help ") {
			topic := strings.TrimSpace(line[6:])
			printHelp(mode, topic)
			continue
		}

		// Handle other dot-commands that forward to the server.
		// These are recognized by the REPL but translated by translateToProtocol.
		if strings.HasPrefix(line, ".") {
			dotLower := strings.ToLower(line)
			isForwarded := dotLower == ".status" ||
				dotLower == ".screen" ||
				dotLower == ".reset" ||
				dotLower == ".warmstart" ||
				dotLower == ".screenshot" ||
				strings.HasPrefix(dotLower, ".screenshot ") ||
				strings.HasPrefix(dotLower, ".state ") ||
				strings.HasPrefix(dotLower, ".boot ")

			if !isForwarded {
				fmt.Fprintf(os.Stderr, "Error: Unknown command: %s\n", line)
				continue
			}
		}

		// Translate REPL command to one or more CLI protocol commands.
		// Some monitor commands (e.g. `g $addr`) expand to multiple
		// sequential protocol commands.
		cliCommands := translateToProtocol(line, mode, atasciiMode)

		// Send each command to the server in order.
		for _, cliCmd := range cliCommands {
			resp, err := client.SendRaw(cliCmd)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error: %v\n", err)
				continue
			}

			// Display response.
			if resp.IsOK() {
				if resp.Data == "" {
					continue
				}

				// Check if the server started an interactive assembly session.
				// Response format: "ASM $XXXX"
				if strings.HasPrefix(resp.Data, "ASM $") {
					if addr, ok := parseHexAddress(resp.Data[4:]); ok {
						inAssemblyMode = true
						assemblyAddress = addr
						continue
					}
				}

				// Handle multi-line responses.
				//
				// GO CONCEPT: Protocol Separator Handling
				// ----------------------------------------
				// The CLI text protocol uses ASCII Record Separator (0x1E)
				// to encode multiple lines in a single response. We replace
				// them with actual newlines for display.
				//
				// Compare with Swift:
				//   output.replacingOccurrences(of: "\u{1E}", with: "\n")
				//
				// Compare with Python:
				//   output.replace("\x1e", "\n")
				output := strings.ReplaceAll(resp.Data, atticprotocol.MultiLineSeparator, "\n")
				fmt.Println(output)
			} else {
				fmt.Fprintf(os.Stderr, "Error: %s\n", resp.Data)
			}
		}
	}
}

File: Python/AtticCLI/attic_cli/repl.py
"""Main REPL loop with mode switching, assembly sub-mode, and tab completion.

This is the interactive heart of the CLI. It reads user input via
prompt_toolkit, dispatches to mode-specific command translators, sends
protocol commands to AtticServer, and displays responses.
"""

from prompt_toolkit import PromptSession
from prompt_toolkit.completion import Completer, Completion, WordCompleter
from prompt_toolkit.document import Document
from prompt_toolkit.formatted_text import HTML
from rich.console import Console

from .cli_client import CLIError, CLISocketClient
from .commands import QUIT, SHUTDOWN, handle_dot_command
from .help import (
    BASIC_HELP,
    DOS_HELP,
    GLOBAL_HELP,
    MONITOR_HELP,
)
from .history import get_history
from .protocol import MULTI_LINE_SEP
from .translator import translate_basic, translate_dos, translate_monitor

console = Console()


def run_repl(client: CLISocketClient) -> None:
    """Run the interactive REPL loop.

    Args:
        client: A connected CLISocketClient instance.
    """
    mode = "basic"          # Default mode (matches Swift CLI)
    current_drive = 1       # Current DOS drive number
    in_assembly = False     # Interactive assembly sub-mode
    assembly_addr = 0       # Next assembly address (hex)

    history = get_history()
    session: PromptSession = PromptSession(history=history)

    try:
        while True:
            # --- Build prompt ---
            if in_assembly:
                prompt = HTML(f"<b>${assembly_addr:04X}</b>: ")
            else:
                prompt = _mode_prompt(mode, current_drive)

            # --- Read input ---
            try:
                completer = _mode_completer(mode, in_assembly)
                line = session.prompt(prompt, completer=completer)
            except EOFError:
                # Ctrl-D: exit
                if in_assembly:
                    _end_assembly(client)
                    in_assembly = False
                    continue
                console.print("\nGoodbye")
                break
            except KeyboardInterrupt:
                # Ctrl-C: cancel current line
                continue

            trimmed = line.strip()

            # --- Interactive assembly sub-mode ---
            if in_assembly:
                if not trimmed or trimmed == ".":
                    result = _end_assembly(client)
                    if result:
                        console.print(result)
                    in_assembly = False
                    continue

                # Send instruction to assembler
                try:
                    response = client.send(f"assemble input {trimmed}")
                    if response.success:
                        payload = response.payload
                        # Response format: "assembled line\x1E$XXXX"
                        if MULTI_LINE_SEP in payload:
                            parts = payload.rsplit(MULTI_LINE_SEP, 1)
                            console.print(parts[0])
                            # Extract next address
                            addr_str = parts[1].strip().lstrip("$")
                            try:
                                assembly_addr = int(addr_str, 16)
                            except ValueError:
                                pass
                        else:
                            console.print(payload)
                    else:
                        console.print(f"[red]Error:[/red] {response.payload}")
                except Exception as exc:
                    console.print(f"[red]Error:[/red] {exc}")
                continue

            # --- Empty line ---
            if not trimmed:
                continue

            # --- Dot-commands (global) ---
            if trimmed.startswith("."):
                result = _handle_dot_with_mode(
                    trimmed, client=client, mode=mode, current_drive=current_drive
                )

                if result is QUIT or result is SHUTDOWN:
                    console.print("Goodbye")
                    break

                if isinstance(result, dict):
                    # Mode change
                    if "mode" in result:
                        mode = result["mode"]
                    if "output" in result and result["output"]:
                        console.print(result["output"])
                elif isinstance(result, str):
                    console.print(result)
                # None means output was already printed (e.g., help)
                continue

            # --- Mode-specific command translation ---
            commands = _translate_for_mode(trimmed, mode)

            for cmd in commands:
                try:
                    response = client.send(cmd)
                    if response.success:
                        payload = response.payload

                        # Check for assembly mode entry
                        if payload.startswith("ASM $"):
                            in_assembly = True
                            addr_str = payload[5:].strip()
                            try:
                                assembly_addr = int(addr_str, 16)
                            except ValueError:
                                assembly_addr = 0
                            console.print(
                                f"[dim]Entering assembly mode at ${assembly_addr:04X}. "
                                f"Type '.' or empty line to exit.[/dim]"
                            )
                            continue

                        # Track drive changes from dos cd
                        if cmd.startswith("dos cd ") and payload.startswith("D"):
                            try:
                                current_drive = int(payload[1:payload.index(":")])
                            except (ValueError, IndexError):
                                pass

                        # Track unmount affecting current drive
                        if cmd.startswith("unmount "):
                            try:
                                drive_num = int(cmd.split()[1])
                                if drive_num == current_drive:
                                    current_drive = 1
                            except (ValueError, IndexError):
                                pass

                        # Display response
                        if payload:
                            if MULTI_LINE_SEP in payload:
                                for part in payload.split(MULTI_LINE_SEP):
                                    console.print(part, highlight=False)
                            else:
                                console.print(payload, highlight=False)
                    else:
                        console.print(f"[red]Error:[/red] {response.payload}")

                except Exception as exc:
                    console.print(f"[red]Error:[/red] {exc}")
                    break

            # --- Drain async events ---
            for event in client.drain_events():
                _display_event(event)

    finally:
        client.disconnect()


def _handle_dot_with_mode(
    line: str,
    *,
    client: CLISocketClient,
    mode: str,
    current_drive: int,
) -> dict | str | object | None:
    """Handle dot-commands, returning mode changes as dicts.

    Returns:
        - dict with "mode" and/or "output" keys for mode changes
        - str for display text
        - QUIT/SHUTDOWN sentinels
        - None for already-printed output
    """
    lower = line.strip().lower()
    parts = line.strip().split(None, 1)
    cmd = parts[0].lower() if parts else ""
    args = parts[1] if len(parts) > 1 else ""

    # Mode switches
    if lower == ".monitor":
        return {"mode": "monitor", "output": "Switched to monitor mode"}
    if lower in (".basic", ".basic atari"):
        return {"mode": "basic", "output": "Switched to BASIC mode"}
    if lower == ".basic turbo":
        return {"mode": "basic_turbo", "output": "Switched to Turbo BASIC mode"}
    if lower == ".dos":
        return {"mode": "dos", "output": "Switched to DOS mode"}

    # Delegate to commands module for everything else
    effective_mode = mode.replace("basic_turbo", "basic")
    return handle_dot_command(
        line,
        client=client,
        mode=effective_mode,
        set_mode=lambda m: None,  # mode switching handled above
    )


def _mode_prompt(mode: str, drive: int) -> HTML:
    """Generate the prompt for the current mode."""
    match mode:
        case "monitor":
            return HTML("<style fg='ansigray'>[monitor]</style> <b>&gt;</b> ")
        case "basic":
            return HTML("<style fg='ansigray'>[basic]</style> <b>&gt;</b> ")
        case "basic_turbo":
            return HTML(
                "<style fg='ansigray'>[basic:turbo]</style> <b>&gt;</b> "
            )
        case "dos":
            return HTML(
                f"<style fg='ansigray'>[dos]</style> <b>D{drive}:&gt;</b> "
            )
        case _:
            return HTML("<b>&gt;</b> ")


class _DotAwareCompleter(Completer):
    """Completer that treats '.' as part of the word being completed.

    WordCompleter splits on '.' so typing '.h<TAB>' fails to match '.help'.
    This completer matches the full text from the last whitespace boundary,
    preserving the leading dot for dot-commands.
    """

    def __init__(self, words: list[str]) -> None:
        self.words = [w.lower() for w in words]

    def get_completions(self, document: Document, complete_event):
        # Get text from the last whitespace to the cursor
        text = document.text_before_cursor
        # Find the last space to get the current "word" including dots
        space_idx = text.rfind(" ")
        if space_idx >= 0:
            current_word = text[space_idx + 1 :]
        else:
            current_word = text
        prefix = current_word.lower()

        for word in self.words:
            if word.startswith(prefix):
                yield Completion(word, start_position=-len(current_word))


def _mode_completer(mode: str, in_assembly: bool) -> _DotAwareCompleter | None:
    """Build a tab completer for the current mode."""
    if in_assembly:
        return None  # No completion in assembly mode

    # Global dot-commands
    global_cmds = [f".{cmd}" for cmd in GLOBAL_HELP]

    match mode:
        case "monitor":
            mode_cmds = list(MONITOR_HELP.keys())
        case "basic" | "basic_turbo":
            mode_cmds = list(BASIC_HELP.keys())
        case "dos":
            mode_cmds = list(DOS_HELP.keys())
        case _:
            mode_cmds = []

    return _DotAwareCompleter(global_cmds + mode_cmds)


def _translate_for_mode(line: str, mode: str) -> list[str]:
    """Translate user input to protocol commands based on current mode."""
    match mode:
        case "monitor":
            return translate_monitor(line)
        case "basic" | "basic_turbo":
            return translate_basic(line)
        case "dos":
            return translate_dos(line)
        case _:
            return [line]


def _end_assembly(client: CLISocketClient) -> str | None:
    """End interactive assembly mode."""
    try:
        response = client.send("assemble end")
        if response.success and response.payload:
            return response.payload
        if not response.success:
            return f"[red]Error:[/red] {response.payload}"
    except Exception as exc:
        return f"[red]Error:[/red] {exc}"
    return None


def _display_event(event) -> None:
    """Display an async event from the server."""
    from rich.panel import Panel

    match event.kind:
        case "breakpoint":
            console.print(Panel(
                f"Breakpoint hit: {event.data}",
                title="BREAKPOINT",
                border_style="yellow",
            ))
        case "stopped":
            console.print(f"[yellow]Program stopped at {event.data}[/yellow]")
        case "error":
            console.print(f"[red]Server error:[/red] {event.data}")
        case _:
            console.print(f"[dim]Event: {event.kind} {event.data}[/dim]")
